rfc/0000-batch-cancellation.rst

```diff
@@ -0,0 +1,538 @@
+database structure is designed for only an entire batch to be
+cancelled.
+
+
+-----------------------------
+Proposed Change Specification
+-----------------------------
+
+**************
+Implementation
+**************
+
+Given the design of our current system, any changes to how
+cancellation works need to adhere to the following constraints:
+
+1. There must be a mechanism in the database that can be updated in an
+   O(1) database operation to indicate a batch has been cancelled.
+2. Queries that scan "Ready" jobs must be fast and respond to a
+   cancellation event instantaneously to avoid unnecessary
+   provisioning of resources.
+3. The data the fair share calculation uses must be updated
+   instantaneously after a cancellation event or be updated within
+   some time window that balances cluster efficiency with code
+   complexity or the fair share algorithm must be redesigned entirely.
+
+The proposed changes to the database are to:
+
+1. Create a new table ``user_inst_coll_resources_by_batch`` that is
+   almost identical to ``user_inst_coll_resources``, but parameterized
+   by an extra field (batch id) and splits the number of ready jobs to
+   "n_ready_always_run_jobs", "n_ready_cancellable_jobs" and likewise
+   for cores and also removes the counts of "n_cancelled_*" fields.
+2. Remove the existing ``batch_inst_coll_cancellable_resources``
+   table.
+3. Remove the `n_cancelled_ready_jobs`, `n_cancelled_running_jobs`
+   etc. columns from the ``user_inst_coll_resources`` table.
+4. Update the jobs after update trigger to not update the old
+   cancellable resources table or the fields from Step 3 and instead
+   update the new ``user_inst_coll_resources_by_batch`` table taking
+   into account the **always_run** flag of the job.
+5. Modify ``cancel_batch`` to just insert a row into the
+   ``batches_cancelled`` table and no longer do the update to the
+   ``user_inst_coll_resources`` table that subtracted the number of
+   newly cancelled resources.
+6. Add a new trigger to the batches table that deletes all rows from
+   ``user_inst_coll_resources_by_batch`` for that batch when the batch
+   state is set to "completed"
+
+
+The proposed changes to the driver are:
+
+1. Change the fair share algorithm in the autoscaler to query the new
+   ``user_inst_coll_resources_by_batch`` table and ignore the values
+   for "n_ready_cancellable_jobs" for batches that have been cancelled
+   when computing the total aggregated values while still accounting
+   for "n_ready_always_run_jobs" for cancelled batches.
+2. Change the fair share algorithm in the scheduler to query the new
+   ``user_inst_coll_resources_by_batch`` table and ignore the values
+   for "n_ready_cancellable_jobs" for batches that have been cancelled
+   when computing the total aggregated values while still accounting
+   for "n_ready_always_run_jobs" for cancelled batches. The fair share
+   proportions per user are cached for 5 seconds at a time in order to
```

Comment:
The fair share proportion changes every scheduling loop, so this feels wrong to cache.